packages/flutter/lib/src/widgets/app.dart

```diff
@@ -969,6 +1063,17 @@ class _WidgetsAppState extends State<WidgetsApp> with WidgetsBindingObserver {
     super.dispose();
   }
 
+  bool get _usesRouter => widget.routerDelegate != null;
+
+  // ROUTER
+
+  RouteNameProvider _routeNameProvider;
+
+  void _updateRouter() {
+    _routeNameProvider = widget.routeNameProvider ??
+      PlatformRouteNameProvider(initialRouteName: _initialRouteName);
```

Comment:
if we create this as a PlatformRouteNameProvider we should call dispose on it too when it goes away (e.g. when it's replaced by widget.routeNameProvider, or when we get disposed)